Task/216 jest config

davemfish opened the pull request:
This PR fixes #216 , it standardizes our approach to jest mocking, setup and teardown.

It adds some global jest config to `package.json` that handles consistent cleanup of mocks before each test. So mock-cleanup is removed from `afterEach` and `afterAll` calls within test files. Please see the changes in `readme.md` for details on how I ended up with that new config.

Also:
* migrates from `fireEvent` to `userEvent`, since that is recommended as a better way to mimic real user interactions.
* updates to jest 27
* Better segregates tests that need a jsdom environment (`renderer` tests) from those that only need node (`main` tests).
* Some other minor changes were inspired by https://kentcdodds.com/blog/avoid-nesting-when-youre-testing

emlys commented on tests/jest.overrides.js:
```diff
@@ -2,11 +2,11 @@ const crypto = require('crypto');
+// console.error = function (message) {
+//   error.apply(console, arguments); // keep default behaviour
+//   throw (message instanceof Error ? message : new Error(message));
+// };
```
Just to double-check, you mean to keep this commented out here, right?

davemfish replied:
I go back and forth on it. But yeah I'm happy enough to leave it commented. It's sometimes useful to uncomment it.

emlys commented on tests/main/main.test.js:
```diff
@@ -115,7 +106,7 @@ describe('extractZipInplace', () => {
   let doneZipping = false;
 
-  beforeEach(() => {
+  beforeEach((done) => {
```
Could you explain what's going on with `done` here? I couldn't find anything about it in the jest docs

davemfish replied:
Here's the docs: https://jestjs.io/docs/asynchronous

And then from https://jestjs.io/docs/setup-teardown:
> beforeEach and afterEach can handle asynchronous code in the same ways that tests can handle asynchronous code - they can either take a done parameter or return a promise.

In this case there is async stuff happening in the setup; we need to wait for the `on('close'...` handler to finish. But there is no Promise we're waiting for, so `done` seemed like a way to control things.

emlys replied:
thanks!

emlys commented on tests/renderer/setuptab.test.js:
```diff
@@ -3,6 +3,7 @@ import React from 'react';
 import {
   createEvent, fireEvent, render, waitFor, within
```
Are there remaining uses of `fireEvent` that could be changed to use `userEvent`?

davemfish replied:
Yeah, the remaining uses are for the drag-and-drop tests. There was no clear way to trigger those events with `userEvent`. Here's a convo about it: https://github.com/testing-library/user-event/issues/440

emlys commented on __mocks__/electron.js:
```diff
@@ -40,6 +40,8 @@ class MockApp extends events.EventEmitter {
       () => os.tmpdir()
     );
     this.setPath = jest.fn();
+    this.getName = jest.fn();
+    this.getVersion = jest.fn();
```
Are these needed because of the switch to `jsdom` environment?

davemfish replied:
Hmm, I can't quite recall why these are needed now and they weren't before. I do remember adding these because a test was triggering these API calls and they're not available outside an electron runtime. So these mocks are necessary, I just can't explain why we didn't need them here earlier.

emlys reviewed (commented):
@davemfish this is great. That readme is super helpful. Just had a few questions, most likely no changes needed.

davemfish commented:
> @davemfish this is great. That readme is super helpful. Just had a few questions, most likely no changes needed.

Thanks @emlys ! As you might be able to tell from readme, I understand this better than I did before, but there are still some minor mysteries. I think I addressed your comments. No changes pushed.

emlys reviewed (approved):
Thanks for the explanations! It makes sense enough to me and we don't need to solve all the mysteries 😄
